### PR TITLE
Improve precision of dirichlet_grad() approximation

### PR DESCRIPTION
**Pull request by fritzo**
This implements @martinjankowiak's improved approximation of reparameterized gradients for the standard Beta and Dirichlet distributions, reducing worst-case relative error from .01 to .001.

See [Derivation](https://github.com/fritzo/notebooks/blob/master/beta-reparameterized.ipynb)

Math reviewed by @martinjankowiak.

### CODE REVIEW

**Comment by fritzo** [comment]
cc @apaszke @gchanan

File: test/test_distributions.py [hunk 1]
```diff
@@ -560,12 +562,12 @@ def test_beta_sample_grad(self):
...
             cdf = scipy.stats.beta.cdf
             pdf = scipy.stats.beta.pdf
-            eps = 0.02 * alpha / (1.0 + np.sqrt(alpha))
+            eps = 0.01 * alpha / (1.0 + np.sqrt(alpha))
             cdf_alpha = (cdf(x, alpha + eps, beta) - cdf(x, alpha - eps, beta)) / (2 * eps)
             cdf_x = pdf(x, alpha, beta)
             expected_grad = -cdf_alpha / cdf_x
             rel_error = np.abs(actual_grad - expected_grad) / (expected_grad + 1e-100)
-            self.assertLess(np.max(rel_error), 0.01,
+            self.assertLess(np.max(rel_error), 0.001,
```
Comment (fritzo) [inline test/test_distributions.py]: Great work @martinjankowiak!

**Comment by martinjankowiak** [comment]
great work @fritzo (where the real credit is due)!!

**Comment by fritzo** [comment]
This conflicts with #4399; we should merge that PR first then I'll update this PR.

**Comment by fritzo** [comment]
Ok I've resolved merge conflicts so this should be ready for code review.

File: aten/src/TH/generic/THTensorMath.c [hunk 2]
```diff
@@ -3492,119 +3492,133 @@ void THTensor_(bhistc)(THTensor *hist, THTensor *tensor, int64_t nbins, real min
...
+  const double term1 = term1_num / term1_den;
+  const double term2 = 0.5f * log(alpha / (total * x));
+  const double term3_num = sqrt(8 * alpha * beta / total);
+  const double term3_den = beta * x + alpha * (x - 1);
+  const double term3 = term3_num / term3_den;
+  const double term4_base = beta * log(beta / (total * (1 - x))) +
+                          alpha * log(alpha / (total * x));
+  const double term4 = pow(term4_base, -1.5f);
+  const double term1234 = term1 + term2 * (term3 + (x < mean ? term4 : -term4));
+  return stirling * prefactor * term1234;
```
Comment (apaszke) [inline aten/src/TH/generic/THTensorMath.c]: Are you using `double` here and `real` everywhere else on purpose? I think it's ok to do all of this math in double

Comment (fritzo) [reply aten/src/TH/generic/THTensorMath.c]: Yes, I'm intentionally using `double` here, as mentioned in the comment. `real` is sufficient precision for all other computations. I'm not sure how we'll port this piece to CUDA.

**Pull request merged**